### PR TITLE
Allow specifying alignment for functions

### LINKED ISSUE #75072: Allow specifying function alignment
I've been recently writing a kernel for the RISC-V architecture, using strictly as much Rust as I can get away with and during which I've run into the need to have certain functions aligned to specific byte alignments (ex. any function placed in `mtvec` MUST be 4 byte aligned), and while I can seemingly accomplish this using `asm!(".align <alignment>")` I think it would make sense for that to be an attribute of the function itself, whether it be limited to `extern "C"` fns or valid on all ABIs. Example of what I am asking for:

```rust
#[no_mangle]
#[repr(align(4))]
extern "C" fn gets_put_in_mtvec() {
    // ... content here ...
}
```

### PR DESCRIPTION
**Pull request by repnop**
Fixes #75072

This allows the user to specify alignment for functions, which can be useful for low level work where functions need to necessarily be aligned to a specific value.

I believe the error cases not covered in the match are caught earlier based on my testing so I had them just return `None`.

### CODE REVIEW

**Comment by repnop** [comment]
Not sure how that submodule got added, but removed :thinking:

**Comment by lcnr** [comment]
nominating for `t-lang` for now, I expect us to land this unstably first if we want to do so, but it does seem desirable to do so imo.

**Comment by joshtriplett** [comment]
This seems reasonable to me, to add on an unstable basis.

I also think many codebases will want to set this for *all* functions using a compiler option (e.g. set in a Cargo profile). But supporting it on individual functions seems reasonable as well.

**Comment by repnop** [comment]
Awesome! I should hopefully have some time soon to do that :)

**Comment by joshtriplett** [comment]
We discussed this in today's @rust-lang/lang meeting, and we're in favor of adding this to nightly, for experimentation, as soon as there's a feature-gate added.

@rfcbot fcp merge
@rfcbot concern feature-gate

**Comment by repnop** [comment]
I think I added all the requests, any other changes that need made?

**Comment by joshtriplett** [comment]
@rfcbot resolve feature-gate

Thank you!

**Comment by Havvy** [comment]
Shouldn't this be going through the RFC process?

**Comment by memoryruins** [comment]
> Shouldn't this be going through the RFC process?

`#[repr(align(x))]` for enums did not require a RFC either in https://github.com/rust-lang/rust/pull/57998, which was introduced as an extension to the repr(align) attribute from [RFC 1358](https://github.com/rust-lang/rfcs/blob/master/text/1358-repr-align.md). It appears to have gone through a similar process and agreement by the lang team here, except this time rfcbot has been involved ^^

**Comment by repnop** [comment]
Anything else before this is merged? :)

File: compiler/rustc_typeck/src/collect.rs [hunk 1]
```diff
@@ -2838,6 +2840,50 @@ fn codegen_fn_attrs(tcx: TyCtxt<'_>, id: DefId) -> CodegenFnAttrs {
...
+                                            Ok(*literal as u32)
+                                        } else {
+                                            Err("larger than 2^29")
+                                        }
+                                    } else {
+                                        Err("not a power of two")
+                                    }
+                                }
+                                _ => Err("not an unsuffixed integer"),
+                            };
```
Comment (lcnr) [inline compiler/rustc_typeck/src/collect.rs:2862]: can you deduplicate this and move it into a function? https://github.com/rust-lang/rust/blob/6523b721c3d8a92cf06aab04dd15ebe7fd5a9fb7/compiler/rustc_attr/src/builtin.rs#L888-L904

File: compiler/rustc_passes/src/check_attr.rs [hunk 2]
```diff
@@ -989,16 +989,36 @@ impl CheckAttrVisitor<'tcx> {
...
                         _ => ("a", "struct, enum, or union"),
                     }
                 }
+                sym::align => {
+                    if let (Target::Fn, true) = (target, !self.tcx.features().fn_align) {
+                        feature_err(
+                            &self.tcx.sess.parse_sess,
+                            sym::fn_align,
+                            hint.span(),
+                            "the attribute `repr(align(...))` on functions is unstable",
```
Comment (lcnr) [inline compiler/rustc_passes/src/check_attr.rs:1012]:
```suggestion
                            "`repr(align)` attributes on functions are unstable",
```
nit: copying the way we write this in https://github.com/rust-lang/rust/blob/6523b721c3d8a92cf06aab04dd15ebe7fd5a9fb7/compiler/rustc_attr/src/builtin.rs#L924

File: compiler/rustc_passes/src/check_attr.rs [hunk 3]
```diff
@@ -989,16 +989,36 @@ impl CheckAttrVisitor<'tcx> {
 
         for hint in &hints {
             let (article, allowed_targets) = match hint.name_or_empty() {
-                _ if !matches!(target, Target::Struct | Target::Enum | Target::Union) => {
-                    ("a", "struct, enum, or union")
+                _ if !matches!(
```
Comment (lcnr) [inline compiler/rustc_passes/src/check_attr.rs:992]: this already existed before, but why do we match eagerly here  :thinking: 

this seems unnecessary and slightly worsens diagnostics. I would prefer us to remove that match arm and
put a `delay_span_bug' in the wildcard arm here.

You don't have to do that in this PR if you want though.

**Review by lcnr — COMMENTED** [review]
a few nits, after that r=me

File: compiler/rustc_passes/src/check_attr.rs (hunk 3 above)
Comment (repnop) [reply compiler/rustc_passes/src/check_attr.rs:992]: figured I may as well, so I added this and fixed the few tests that failed, except for one: https://github.com/rust-lang/rust/blob/36f1f04f18b89ba4a999bcfd6584663fd6fc1c5d/src/test/ui/issues/issue-43988.rs#L14-L22 which has two errors not longer printed, I'm assuming because `delay_span_bug` stuff isn't emitted if there are other errors before that? is there a particular way I should resolve that?

Comment (repnop) [reply compiler/rustc_passes/src/check_attr.rs:992]: (friendly poke :D)

Comment (lcnr) [reply compiler/rustc_passes/src/check_attr.rs:992]: what happens if we change the last match arm to something like
```rust
_ => struct_span_err!(
                self.tcx.sess,
                hint.span(),
                E0552,
                "unrecognized representation hint"
            )
            .emit();
```
you might have to change `ReprOptions::new` to a `delay_span_bug` instead.

**Review by lcnr — COMMENTED** [review]
sorry for taking so long to get back to you, wanted to look at the code itself to figure out why the test case changed and didn't find the time to do so.

File: compiler/rustc_passes/src/check_attr.rs (hunk 3 above)
Comment (repnop) [reply compiler/rustc_passes/src/check_attr.rs:992]: I'll give that a go in a couple hours and report back :)

Comment (repnop) [reply compiler/rustc_passes/src/check_attr.rs:992]: changed the final match arm to
```rust
_ => {
    struct_span_err!(
        self.tcx.sess,
        hint.span(),
        E0552,
        "unrecognized representation hint"
    )
    .emit();
    continue;
}
```
which resulted in the following other test failure:
```
error[E0565]: meta item in `repr` must be an identifier
  --> /home/repnop/dev/rust/rust/src/test/ui/error-codes/E0565.rs:2:8
   |
LL | #[repr("C")] //~ ERROR E0565
   |        ^^^

error[E0552]: unrecognized representation hint
  --> /home/repnop/dev/rust/rust/src/test/ui/error-codes/E0565.rs:2:8
   |
LL | #[repr("C")] //~ ERROR E0565
   |        ^^^

error[E0565]: meta item in `repr` must be an identifier
  --> /home/repnop/dev/rust/rust/src/test/ui/error-codes/E0565.rs:2:8
   |
LL | #[repr("C")] //~ ERROR E0565
   |        ^^^

error: aborting due to 3 previous errors

Some errors have detailed explanations: E0552, E0565.
For more information about an error, try `rustc --explain E0552`.
```

the output for `issue-43988.rs` now looks like:
```
error: malformed `repr` attribute input
  --> /home/repnop/dev/rust/rust/src/test/ui/issues/issue-43988.rs:24:5
   |
LL |     #[repr]
   |     ^^^^^^^ help: must be of the form: `#[repr(C)]`

error: malformed `repr` attribute input
  --> /home/repnop/dev/rust/rust/src/test/ui/issues/issue-43988.rs:34:14
   |
LL |     let _z = #[repr] 1;
   |              ^^^^^^^ help: must be of the form: `#[repr(C)]`

error[E0518]: attribute should be applied to function or closure
  --> /home/repnop/dev/rust/rust/src/test/ui/issues/issue-43988.rs:5:5
   |
LL |     #[inline]
   |     ^^^^^^^^^
LL |     let _a = 4;
   |     ----------- not a function or closure

error[E0518]: attribute should be applied to function or closure
  --> /home/repnop/dev/rust/rust/src/test/ui/issues/issue-43988.rs:10:5
   |
LL |     #[inline(XYZ)]
   |     ^^^^^^^^^^^^^^
LL |     let _b = 4;
   |     ----------- not a function or closure

error[E0552]: unrecognized representation hint
  --> /home/repnop/dev/rust/rust/src/test/ui/issues/issue-43988.rs:14:12
   |
LL |     #[repr(nothing)]
   |            ^^^^^^^

error[E0552]: unrecognized representation hint
  --> /home/repnop/dev/rust/rust/src/test/ui/issues/issue-43988.rs:18:12
   |
LL |     #[repr(something_not_real)]
   |            ^^^^^^^^^^^^^^^^^^

error[E0518]: attribute should be applied to function or closure
  --> /home/repnop/dev/rust/rust/src/test/ui/issues/issue-43988.rs:30:5
   |
LL |     #[inline(ABC)]
   |     ^^^^^^^^^^^^^^
LL |     foo();
   |     ----- not a function or closure

error: aborting due to 7 previous errors

Some errors have detailed explanations: E0518, E0552.
For more information about an error, try `rustc --explain E0518`.
```
which is a definite win, I think.

as for the change to `ReprOptions::new`, I don't see anything that emits errors in the function, so I'm assuming nothing needs adjusted in it.

Comment (lcnr) [reply compiler/rustc_passes/src/check_attr.rs:992]: ah yeah, meant `fn find_repr_attrs` instead of `ReprOptions::new`. If you move the error reporting from there to `check_attrs` this should be a cleaner solution.

So changing the following to a delay_span_bug:
```rust
                if !recognised {
                    // Not a word we recognize
                    struct_span_err!(
                        diagnostic,
                        item.span(),
                        E0552,
                        "unrecognized representation hint"
                    )
                    .emit();
                }
```

and completely moving the `if !item.is_meta_item()`, should make this a lot cleaner i think

Comment (repnop) [reply compiler/rustc_passes/src/check_attr.rs:992]: just to make sure I'm understanding correctly: do you want me to move all of the error reporting from `find_repr_attrs` to `check_attrs` or just those specific instances? :thinking:

Comment (lcnr) [reply compiler/rustc_passes/src/check_attr.rs:992]: i think it's easier to just move those specific instances. If you're able to cleanly move all the error reporting to here that's also fine for me

Comment (repnop) [reply compiler/rustc_passes/src/check_attr.rs]: okay I _think_ I got it

File: src/test/ui/issues/issue-43988.rs [hunk 4]
```diff
@@ -11,16 +11,6 @@ fn main() {
     let _b = 4;
     //~^^ ERROR attribute should be applied to function or closure
 
-    #[repr(nothing)]
```
Comment (lcnr) [inline src/test/ui/issues/issue-43988.rs:14]: why did you remove these 2 reprs?

**Review by lcnr — APPROVED** [review]

File: src/test/ui/issues/issue-43988.rs (hunk 4 above)
Comment (repnop) [reply src/test/ui/issues/issue-43988.rs:14]: whoops! I think that may have been a leftover of me trying to figure out something, let me try adding them back here.

**Comment by lcnr** [comment]
if ci passes, please squash your commits

after that r=me

@bors delegate+

**Comment by RalfJung** [comment]
What are the language-level effects of this? Is this mainly a hint for how to put the compiled binary code together, or is his somehow observable inside Rust?

**Comment by repnop** [comment]
> Is this mainly a hint for how to put the compiled binary code together
yes, its not intended to introduce any visible effects from within Rust, but to allow telling LLVM that the function needs to be aligned to a specific boundary for situations in which that's necessary and you can't rely on the native alignment placing it at an appropriate address (e.g. so it can be inserted into a hardware register, in my case)

> by adding new kinds of UB
I think the only thing that may have been glossed over in this PR as it stands right now is being able to under-align functions, which I'll need to investigate to see if we need to add some additional checks to the `#[repr(align(...))]` as to not allow that to happen, or if LLVM will ignore the alignment hint and force it up to the native alignment since it would fulfill the request anyways (which seems like it may be the case? at least with some limited test in the playground)
